sway-core/src/decl_engine/id.rs

```diff
@@ -1,71 +1,228 @@
-        &self.0
+    pub(crate) fn replace_id(&mut self, index: Self) {
+        self.0 = index.0;
     }
 }
 
 #[allow(clippy::from_over_into)]
-impl Into<usize> for DeclId {
+impl<T> Into<usize> for DeclId<T> {
     fn into(self) -> usize {
         self.0
     }
 }
 
-impl From<&DeclId> for DeclId {
-    fn from(value: &DeclId) -> Self {
+impl<T> From<&DeclId<T>> for DeclId<T> {
+    fn from(value: &DeclId<T>) -> Self {
         DeclId::new(value.0)
     }
 }
 
-impl From<&mut DeclId> for DeclId {
-    fn from(value: &mut DeclId) -> Self {
+impl<T> From<&mut DeclId<T>> for DeclId<T> {
+    fn from(value: &mut DeclId<T>) -> Self {
         DeclId::new(value.0)
     }
 }
 
-impl From<&DeclRef> for DeclId {
-    fn from(value: &DeclRef) -> Self {
+impl<T> From<&DeclRef<DeclId<T>>> for DeclId<T> {
+    fn from(value: &DeclRef<DeclId<T>>) -> Self {
         value.id
     }
 }
 
-impl From<&mut DeclRef> for DeclId {
-    fn from(value: &mut DeclRef) -> Self {
+impl<T> From<&mut DeclRef<DeclId<T>>> for DeclId<T> {
+    fn from(value: &mut DeclRef<DeclId<T>>) -> Self {
         value.id
     }
 }
 
-impl SubstTypes for DeclId {
+impl SubstTypes for DeclId<TyFunctionDeclaration> {
+    fn subst_inner(&mut self, type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let mut decl = decl_engine.get(*self);
+        decl.subst(type_mapping, engines);
+        decl_engine.replace(*self, decl);
+    }
+}
+impl SubstTypes for DeclId<TyTraitDeclaration> {
+    fn subst_inner(&mut self, type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let mut decl = decl_engine.get(*self);
+        decl.subst(type_mapping, engines);
+        decl_engine.replace(*self, decl);
+    }
+}
+impl SubstTypes for DeclId<TyTraitFn> {
+    fn subst_inner(&mut self, type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let mut decl = decl_engine.get(*self);
+        decl.subst(type_mapping, engines);
+        decl_engine.replace(*self, decl);
+    }
+}
+impl SubstTypes for DeclId<TyImplTrait> {
+    fn subst_inner(&mut self, type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let mut decl = decl_engine.get(*self);
+        decl.subst(type_mapping, engines);
+        decl_engine.replace(*self, decl);
+    }
+}
+impl SubstTypes for DeclId<TyStructDeclaration> {
+    fn subst_inner(&mut self, type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let mut decl = decl_engine.get(*self);
+        decl.subst(type_mapping, engines);
+        decl_engine.replace(*self, decl);
+    }
+}
+impl SubstTypes for DeclId<TyStorageDeclaration> {
+    fn subst_inner(&mut self, _type_mapping: &TypeSubstMap, engines: Engines<'_>) {
+        let decl_engine = engines.de();
+        let decl = decl_engine.get(*self);
+        // decl.subst(type_mapping, engines);
```

Comment:
this line is included in the above impls but commented out for this one and a few below, is this intentional?

Reply:
Yes, it's very much intentional but I should add a comment to clarify.

Reply:
Turns out the function can be gotten rid of altogether.